### ISSUE TITLE
Gas completeness (EIP-1559, base fee oscillation, gas premiums, block gas limit)

### ISSUE THREAD

**Issue by maciejwitowski** [consensus-shipyard/ipc#925]
From @raulk:

TL;DR: The high level goal of this project is to complete the gas system in IPC. We propose to follow the EIP-1559 model by default, but providing customizability for the oscillation function / initial base fee / limits / more. This epic also fixes gas premium handling and burns.

## Status quo in IPC

- Base fee configured at genesis is static; it does not oscillate based on block utilization.
- IPC has no gas block limits yet (but it does have a configurable physical block size limit with a default of 21MiB, see below). Hence it cannot calculate compute utilization because it has no concept of limited capacity.
- When executing a tx, the subnet burns the network fee by transferring it to the burnt funds actor (f099).
- When executing a tx, ref-fvm credits premiums / priority fees to the reward actor (f02), but the client doesn't disburse it to the block proposer like it happens in Filecoin:
	- At the end of the block, Filecoin clients execute an `AwardBlockRewards` implicit transaction to withdraw the cumulative premiums + block reward to the miner.
- We have evidence that `eth_maxPriorityFeePerGas` and/or `eth_feeHistory` are miscalculating something and causing clients to propose extremely spiky priority gas fees.

## Proposed solution

- Investigate and fix `eth_maxPriorityFeePerGas` and/or `eth_feeHistory` to prevent clients from choosing spiky priority fees.
- Send the gas premium to the block proposer.
	- This may require linking a validator pubkey with an account.
	- Evaluate if we maintain or kill the reward actor indirection. This exists in Filecoin to dispense block rewards via inflation subject to locking/vesting, but IPC may choose a simpler path. On the other hand, the abstraction is worth preserving as an extension/customization point.
	- One way or another, I'm pretty sure we'll need to override the default reward actor with our own.
- Introduce a configurable _soft_ block gas limit (EIP-1559 target) and a _hard_ gas limit, as well as a physical maximum block size (ensure we have a maximum transaction size).
	- These can go in the chain metadata actor and be initialized at genesis, then be adjusted later by the subnet owner via on-chain transactions (without requiring an upgrade).
	- Ensure our block production logic only produces valid blocks.
	- Ensure our block validation only accepts valid blocks.
- Introduce a configurable base fee oscillation function, defaulting to EIP-1559.
- Reconcile burns (f099) with circulating supply at the parent chain. Perhaps we _physically_ burn these amounts and inform the parent of the burnt delta in checkpoints. The parent will need to apply the burn locally on the supply source (native/token), and in turn inform its parent (this mechanism should work for L3+).
- Provide a migration path for existing users (including recommendations to deal with funds held at f02).

**Comment by raulk** [comment]
An additional problem with premiums is that explorers like Blockscout assume the premium has gone to the miner. And the miner address is incorrectly calculated as the hash of the proposer's Tendermint ID. See:

https://github.com/consensus-shipyard/ipc/blob/ebb743f42882efa7af8f4a31970e9c854893238b/fendermint/eth/api/src/conv/from_tm.rs#L136

![image](https://github.com/consensus-shipyard/ipc/assets/1101242/874eb501-d270-4331-b461-87094e9630b6)

![image](https://github.com/consensus-shipyard/ipc/assets/1101242/f60ed1cc-913a-4abf-8806-e79281dff994)

**Comment by raulk** [comment]
# Sending a single transaction spikes eth_maxPriorityFeePerGas

## Report

Fluence reported transactions with exorbitant fees: [1](https://blockscout-dar.fluence.dev/tx/0x315861be0ceabcc380e35303ae1ae6423c72e3969b0d6559893c14366fddbcf8) [2](https://blockscout-dar.fluence.dev/tx/0x0191a917e196221f7ee07ad37383cd017cf1261e19d0c3e05338d4ba1b67fad1).

Explorer: [https://blockscout-dar.fluence.dev/](https://blockscout-dar.fluence.dev/)

Indeed, these transactions carry gigantic maxPriorityFeePerGas:

```json
⟩ cast rpc --rpc-url https://ipc-dar.fluence.dev eth_getTransactionByHash 0xd29f35c3badab8b6b29f1a994fd4a581137e0907c30c804f0f1c9ba4b4d9942f
{"hash":"0xd29f35c3badab8b6b29f1a994fd4a581137e0907c30c804f0f1c9ba4b4d9942f","nonce":"0x270","blockHash":"0x238db68d992a0a19c54d4429c4d884259db603bbcbfb526cff05d5d18b906ef4","blockNumber":"0x15922b","transactionIndex":"0x0","from":"0x57a0e00fe79f7cf5a01f6d849a59cdc05477d056","to":"0x6fba3f1f130e810f5c95ea4613ec5bd42439fde6","value":"0x0","gasPrice":"0x2540be4465","gas":"0x91bdcc5","input":"0x4ece5685cb722e8d76c115119e2f5a6443ab9b55ba31b2c1e0d56be567a614758db4e74f6ef5dc88c7a6825ef244accc9ac7fcb63bbd66946e5d437c5d2128181f5992130000046df06164406b8a29b15c1cd301e42319701e04b555e431a182ab40909b","v":"0x0","r":"0x9cd77c467beae0b649e6546135a58efbb466d3d5d7ed748e47e04aeba890dccf","s":"0x6b99ad51d45082607b2c5257335fd85cb3e57bd05f94a338d236c6897d2028a8","type":"0x2","accessList":[],"maxPriorityFeePerGas":"0x12a05f2000","maxFeePerGas":"0x12a05f2465","chainId":"0xc8607b57bf789"}
```

They simply pull eth_maxPriorityFeePerGas from our endpoint [\[link\]](https://github.com/fluencelabs/nox/blob/1bbc0d7bfa335fa19ecb43f1b5f199dd0055753b/crates/chain-connector/src/connector.rs#L176):

```
let base_fee_per_gas = self.get_base_fee_per_gas().await?;
let gas_limit = self.estimate_gas_limit(&data, to).await?;
let max_priority_fee_per_gas = self.max_priority_fee_per_gas().await?;
let increase = base_fee_per_gas.div(BASE_FEE_PREMIUM_DIVIDER);
let base_fee = base_fee_per_gas.checked_add(increase).unwrap_or(U256::MAX);
// (base fee + priority fee).
let max_fee_per_gas = base_fee + max_priority_fee_per_gas;
```

## Live test

I asked them to submit a transaction real-time while I recorded the output of eth_maxPriorityFeePerGas. They sent this:

[https://blockscout-dar.fluence.dev/tx/0xbaf11b61d31e33486fa298e069159da4de8b28191267ed55ebaaab6e6c481843](https://blockscout-dar.fluence.dev/tx/0xbaf11b61d31e33486fa298e069159da4de8b28191267ed55ebaaab6e6c481843)

And indeed, I observed a 112x increase in the suggested priority fee, even though the chain is not saturated:

```
⟩ cast rpc --rpc-url https://ipc-dar.fluence.dev estimate_priority_fee
"0x30d40"
⟩ cast rpc --rpc-url https://ipc-dar.fluence.dev eth_maxPriorityFeePerGas
"0x30d40"
⟩ cast rpc --rpc-url https://ipc-dar.fluence.dev eth_maxPriorityFeePerGas
"0xc35000"
⟩ cast rpc --rpc-url https://ipc-dar.fluence.dev eth_maxPriorityFeePerGas
"0x10c8e00"
⟩ cast rpc --rpc-url https://ipc-dar.fluence.dev eth_maxPriorityFeePerGas
"0x10c8e00"
⟩ cast rpc --rpc-url https://ipc-dar.fluence.dev eth_maxPriorityFeePerGas
"0x155cc00"
⟩ cast rpc --rpc-url https://ipc-dar.fluence.dev eth_maxPriorityFeePerGas
"0xaae600"
⟩ cast rpc --rpc-url https://ipc-dar.fluence.dev eth_maxPriorityFeePerGas
"0x30d40"
```

I think something's broken in eth_maxPriorityFeePerGas, it should not be this sensitive.

## Configuration

```
# Any over-estimation to apply on top of the estimate returned by the API.
gas_overestimation_rate = 2

# FVM configuration
[fvm]
# Overestimation rate applied to gas estimations to ensure that the
# message goes through
gas_overestimation_rate = 1.25
# Gas search step increase used to find the optimal gas limit.
# It determines how fine-grained we want the gas estimation to be.
gas_search_step = 1.25
# Indicate whether transactions should be fully executed during the checks performed
# when they are added to the mempool, or just the most basic ones are performed.
#
# Enabling this option is required to fully support "pending" queries in the Ethereum API,
# otherwise only the nonces and balances are projected into a partial state.
exec_in_check = true

# Gas fee used when broadcasting transactions.
# TODO: Configure a value once validators are charged for the "miner penalty".
gas_fee_cap = 0
# Gas premium used when broadcasting transactions.
gas_premium = 0

# Ethereum API facade
[eth]
# Maximum time allowed between polls for filter changes, in seconds, before the subscription is canceled.
filter_timeout = 300
# Maximum number of entries in the LRU caches.
cache_capacity = 1000000

[eth.gas]
# Minimum gas premium returned by the API in `eth_maxPriorityFeePerGas`, in atto.
min_gas_premium = 100000
# Number of blocks used for the computation of the premium in `eth_maxPriorityFeePerGas`
# Default: Lotus uses only 2 epochs to compupte the premium, but they compute the
# median over (on average) 10 blocks, 5 per epoch.
num_blocks_max_prio_fee = 10
# Maximum size of the histogram for `eth_feeHistory`
max_fee_hist_size = 1024
```

**Comment by cryptoAtwill** [comment]
For `eth_maxPriorityFeePerGas` and/or `eth_feeHistory`, we are referencing the existing eth or fil-evm implementation or actually they are aligned?

**Comment by raulk** [comment]
# Implement EIP-1559 base fee oscillation

## Context

Our base fee is currently static: set at genesis and unchanging.

The base fee is a concept that Filecoin borrowed from EIP-1559 (fun fact: Filecoin shipped this behaviour before Ethereum!). It's dynamic and deterministic. It varies every block according to a predefined formula that correlates with block occupancy and, therefore, chain congestion.

Because of that, it's trivial for clients to calculate the base fee cap to guarantee transaction inclusion within N blocks (usually by assuming 100% occupancy rate from now and then).

Note that we currently have no block gas limits ([ENG-582](https://linear.app/interplanetary-consensus/issue/ENG-582/add-block-gas-limits)), so depending how we implement base fee oscillation, we may need to add those first.

## Chain metadata actor

To enable full customizability, we should add the base fee calculation logic to the Chain Metadata actor, instead of to the client codebase. This prepares better us for "yellow runtime", allowing subnet operators to define this logic and have it shipped to validators as Wasm bundles. We'll need to figure out which stats we pass in at the end of the block to enable a general calculation.

Every block proposer includes the base fee update transaction as an implicit transaction to the end of the block. Blocks without this transaction are deemed invalid.

## FVM execution

When constructing the machine, we read the next base fee from the Chain metadata actor and feed it to the FVM Machine.

**Comment by raulk** [comment]
> For `eth_maxPriorityFeePerGas` and/or `eth_feeHistory`, we are referencing the existing eth or fil-evm implementation or actually they are aligned?

@cryptoAtwill I think you're asking if the Fendermint implementations were "inspired" by Lotus or by Ethereum clients? I would imagine the former. We do seem to inherit the FVM block gas limit for calculations:

https://github.com/consensus-shipyard/ipc/blob/ebb743f42882efa7af8f4a31970e9c854893238b/fendermint/eth/api/src/apis/eth.rs#L110

But this is moot, because the maximum block gas is "unbounded": https://github.com/consensus-shipyard/ipc/blob/fdcca99f6f97b35927d2336cff3f45d573a358ed/fendermint/app/src/cmd/genesis.rs#L230

OTOH, we do seem to enforce a max block byte size: https://github.com/consensus-shipyard/ipc/blob/fdcca99f6f97b35927d2336cff3f45d573a358ed/fendermint/app/src/cmd/genesis.rs#L229

But it's 21MiB by default. Not sure why that value is chosen 🤷  Seems too large a default value to me, and probably misaligned with other defaults (3s commit timeout ~= block time)

**Comment by maciejwitowski** [comment]
These 2 parts aren't clear to me - can you add more details?

> This may require linking a validator pubkey with an account.

> Perhaps we physically burn these amounts

**Comment by cryptoAtwill** [comment]
@raulk Regarding the priority spike, there is a small mistake in `premium` calculation at this [line](https://github.com/consensus-shipyard/ipc/blob/918229fd6625ff28b105ae4604ba65497884b2d9/fendermint/eth/api/src/gas/mod.rs#L71).
Instead of:
```rust
if prev2 != TokenAmount::zero() {
    premium += &prev2;
    premium.div_ceil(BigInt::from(2));
}
premium
```
It should be:
```rust
if prev2 != TokenAmount::zero() {
    premium += &prev2;
    premium.div_ceil(BigInt::from(2));
} else {
    premium
}
```

**Comment by raulk** [comment]
@cryptoAtwill hm, are you sure? It looks to me that the current code and what you propose are logically equivalent, when put in the original context?

**Comment by cryptoAtwill** [comment]
@raulk Yeah, my bad, I forgot to delete a `;`.

**Comment by cryptoAtwill** [comment]
From initial review, I think we could consider adding the following to the premium gas estimation:
1. Adding a gas cache, so that we dont have to recalculate the estimation every single request
2. Use `gas_limit` instead of `gas_used` to align with lotus. If we want to create some flexibility, then maybe a `GasLayer` or something like that would help.
3. Allow not just `median` but a percentile of the historical premiums.

**Comment by cryptoAtwill** [comment]
One proposal is to include a `GasLayer` trait that handles the gas related operations. The trait is defined as:
```rust
type Gas = u128;
type State = BlockchainState;

trait GasLayer {
  fn block_gas_limit(&self, state: &State) -> Gas;
  
  fn update_block_gas_limit(&self, state: &mut State, new_limit: Gas);  
}
```
We can extend the above trait with more gas related methods.

The default implementation of `GasLayer` is calling into a custom actor that stores the current gas parameters in the state. We can also include a simple in memory cache as well.
![image](https://github.com/user-attachments/assets/7fd678a4-1bcb-456a-8f19-81f9ef87e857)
